awscli/shorthand.py

```diff
@@ -0,0 +1,264 @@
+    _DOUBLE_QUOTED = _NamedRegex('double quoted', r'"(?:\\\\|\\"|[^"])*"')
+    _FIRST_VALUE = _NamedRegex('first',
+                               u'[\!\#-&\(-\+\--\<\>-Z\\\\-z\u007c-\uffff]'
+                               u'[\!\#-&\(-\+\--\\\\\^-\|~-\uffff]*')
+    _SECOND_VALUE = _NamedRegex('second',
+                                u'[\!\#-&\(-\+\--\<\>-Z\\\\-z\u007c-\uffff]'
+                                u'[\!\#-&\(-\+\--\<\>-\uffff]*')
+
+    def __init__(self):
+        self._tokens = []
+
+    def parse(self, value):
+        self._input_value = value
+        self._index = 0
+        return self._parameter()
+
+    def _parameter(self):
+        # parameter = keyval *("," keyval)
+        params = {}
+        params.update(self._keyval())
+        while self._index < len(self._input_value):
+            self._consume_whitespace()
+            self._expect(',')
+            self._consume_whitespace()
+            params.update(self._keyval())
+        return params
+
+    def _keyval(self):
+        # keyval = key "=" [values]
+        key = self._key()
+        self._consume_whitespace()
+        self._expect('=')
+        self._consume_whitespace()
+        values = self._values()
+        return {key: values}
+
+    def _key(self):
+        # key = 1*(alpha / %x30-39)  ; [a-zA-Z0-9]
+        valid_chars = string.ascii_letters + string.digits
+        start = self._index
+        while not self._at_eof():
+            if self._current() not in valid_chars:
+                break
+            self._index += 1
+        return self._input_value[start:self._index]
+
+    def _values(self):
+        # values = csv-list / explicit-list / hash-literal
+        if self._at_eof():
+            return None
+        elif self._current() == '[':
+            return self._explicit_list()
+        elif self._current() == '{':
+            return self._hash_literal()
+        else:
+            return self._csv_list()
+
+    def _csv_list(self):
+        first_value = self._first_value()
+        self._consume_whitespace()
+        if self._at_eof() or self._input_value[self._index] != ',':
+            return first_value
+        self._consume_whitespace()
+        self._expect(',')
+        self._consume_whitespace()
+        csv_list = [first_value]
+        while True:
+            try:
+                current = self._second_value()
+                if current is None:
+                    break
+                self._consume_whitespace()
+                if self._at_eof():
+                    csv_list.append(current)
+                    break
+                self._expect(',')
+                self._consume_whitespace()
+                csv_list.append(current)
+            except ShorthandParseError:
+                # Backtrack to the previous comma.
+                self._backtrack_to(',')
+                break
+        if len(csv_list) == 1:
+            return first_value
```

Comment:
Yeah exactly.

So maybe it makes sense for csv list to always return a list, and then have the walker that compares against the model do the `[a] -> a` conversion instead of the walker doing a `a -> [a]` conversion.

I think the only reason I did it this way is that it's much more common to have key val pairs where the model actually has the value being a scalar value, but I think from an implementation standpoint it makes more sense to always have csv_list return a list.  That means we parse key=<anything> such that <anything> is always a list.


Thoughts?